### ISSUE TITLE
Signal/Slot should use an object when calling a slot

### ISSUE THREAD

**Issue by kdambekalns** [neos/flow-development-collection#1003]
Right now the arguments passed to a signal are inherently "unstable", in the sense that what is passed is not strictly defined. So just switching on the "pass signal information" flag in wiring could have unintended side-effects.

We (@kitsunet and I) concluded: it would be good to pass an object instead, which could carry any arguments like before but have a defined API to fetch those as well as further data about the signal.

This could be added in a b/c way by checking the type of the (sole) argument expected by the slot.

**Comment by robertlemke** [comment]
The arguments of a **signal** (that is, the PHP method whose name starts with `emit`) requires a defined set of arguments, for example:

```php
    /**
     * Emits a signal that the bootstrap finished and is shutting down.
     *
     * @param string $runLevel
     * @return void
     * @Flow\Signal
     */
    protected function emitBootstrapShuttingDown($runLevel)
    {
        $this->earlyInstances[Dispatcher::class]->dispatch(__CLASS__, 'bootstrapShuttingDown', [$runLevel]);
    }
```

The arguments passed to a **slot** method are the same which were defined in the signal method, but a slot may choose to not use them (or not define them in its method signature). While it's best practice to declare all the arguments of a signal, there's really no need to do so, if the slot doesn't use them.

Can you elaborate on the actual downside, maybe based on an example?

**Comment by kitsunet** [comment]
`Dispatcher::connect($signalClassName, $signalName, $slotClassNameOrObject, $slotMethodName = '', $passSignalInformation = true)`

The  `passSignalInformation` currently means an additional argument (containing the signal class/method as string) is added to the slot call. But that is rather indeterministic especially with optional arguments.

An object similar to what the JoinPoint in AOP does with an defined interface to get arguments and slot information would therefore be beneficial to make that explicit.

**Comment by bwaidelich** [comment]
Could you explain a little how you expect the behavior to be changed?

e.g. do you suggest, given the following signal:

```php
class SomeClass {
  public function emitSomeSignal(int $number, string $string): void;
}
```

...that the corresponding slot would have to change from:

```php
$dispatcher->connect(SomeClass::class, 'someSignal', function(int $number, string $string, string $signalInformation) { ... });
```

to 

```php
$dispatcher->connect(SomeClass::class, 'someSignal', function(int $number, string $string, SignalInformation $signalInformation) { ... });
```

or to sth like

```php
$dispatcher->connect(SomeClass::class, 'someSignal', function(SignalInformationAndArguments $args) { ... });
```

?

**Comment by bwaidelich** [comment]
*ping*

**Comment by bwaidelich** [comment]
@kdambekalns @robertlemke @kitsunet _ping²_ ;)

**Comment by kitsunet** [comment]
I rather meant that we define a singular interface for connecting, as in your latter example, that way it's perfectly deterministic and you can IMHO react better to optional arguments, we might even be able to pass more detailed signal information as well. Just like the AOP joinPoint contains information about the join point and the arguments of the original call.

**Comment by kdambekalns** [comment]
That being said, this is an issue I'd be really interested in to implement…

**Comment by sorenmalling** [comment]
> That being said, this is an issue I'd be really interested in to implement…

I would love to learn how this will be implemented 🤓

**Comment by bwaidelich** [comment]
Great, thanks for clarifying. I'll add the issue to the project board. But IMO this will need a code migration or it will be a hell of a nasty breaking change :)

**Comment by robertlemke** [comment]
> But IMO this will need a code migration or it will be a hell of a nasty breaking change :)

If the change is mostly about the `connect` part, can't we introduce a new method (for example `wire`) in order to stay backwards-compatible, _and_ provide a code migration so that you can migrate to the new way automatically?

**Comment by kitsunet** [comment]
Yep, exactly that would be my thought @robertlemke a new method and staying backwards compatible for another major version or two, it's just not nice to use in some instances, it's not that we urgently NEED to get rid of it.

**Comment by bwaidelich** [comment]
So the way it is implemented with #2216 reflects the _first_ option I described above:
```php
$dispatcher->connect(SomeClass::class, 'someSignal', function(int $number, string $string, string $signalInformation) { ... });
```
would change to
```php
$dispatcher->connect(SomeClass::class, 'someSignal', function(int $number, string $string, SignalInformation $signalInformation) { ... });
```
(the old version would still work due to some reflection trickery).


I welcome that this is not a breaking change now. But I wonder if it really solves the core issue:

> Right now the arguments passed to a signal are inherently "unstable" [...]

In fact it makes it even a little more unstable since the arguments now depend on the consumer.

@kdambekalns any thoughts on why you didn't follow @robertlemke's suggestion to support something along the lines of:
```php
$dispatcher->wire(SomeClass::class, 'someSignal', function(SignalInformationAndArguments $args) { ... });
```
?

**Comment by kdambekalns** [comment]
I did not want to have another method, TBH. My goal was to avoid any potential mishaps with passing of signal information, but… well. Now `wire()` exists, which allows for nice slots. And `connect()` stays, as it allows to use arbitrary methods as slots, which can be handy.

The problem of signal information being passed by default to methods that may accept a variable number of arguments and thus be confused still exists – for that, the default for `$passSignalInformation` would need to be flipped. To raise awareness, I added a note to the documentation… (see https://github.com/neos/flow-development-collection/pull/2217, too)
